### DISCUSSION TITLE
v4.14.13 - New synchronous JavaScript function getLogger() + deprecated async function createLogger()

### DISCUSSION THREAD

**Discussion by jongpie** [jongpie/NebulaLogger#779] [Announcements]
This release is focused on several enhancements & bugfixes for logging in JavaScript (aura & lightning web components). Thanks to @jamessimone for reviewing & discussing the changes for this release, and thanks to @srikanth3107 for reporting issue #776!

# Core Unlocked Package Changes

## New `getLogger()` function

Partially fixed #728 by adding a new function `getLogger()` in `logger` LWC that can be called synchronously, and deprecated the async function `createLogger()`
- This simplifies how developers use the `logger` LWC, and avoids some lingering JavaScript (JS) stack trace issues that occur in `async` functions
- The function `createLogger()` is still supported & functional, but it's now considered deprecated since it requires using `await` (which is no longer necessary with `getLogger()`)

> [!NOTE]
> This does not truly "fix" or improve the stack trace parsing used in `async` functions, and `async` functions will continue to have inaccurate function names reported in some browsers (typically, `eval` is listed as the function name). This is a ongoing challenge in JavaScript due to what info is/isn't available in stack traces in some browsers (most notably, webkit browsers like Chrome & Edge). 
>
> But with the new `getLogger()` function, Nebula Logger no longer _requires_ the use of `async` / `await` - and often the only reason developers were using `async` / `await` was to be able to use Nebula Logger. In these situations, developers can eliminate `async` / `await`, and the resulting log entries will have accurate JS function names.

### Example: Old `createLogger()` Usage

Previously, JS developers had to use `async`, `await`, and truthy checks to ensure that Nebula Logger's LWC had finished being loaded. This resulted in more complexity for developers, who just want to be able to log some stuff 😢 

```javascript
import { createLogger } from 'c/logger';

export default class LoggerDemo extends LightningElement {
  logger;

  // Some lifecycle event (typically connectedCallback()) has to run async & await createLogger()
  async connectedCallback() {
    this.logger = await createLogger();
    this.logger.info('Hello, world');
    this.logger.saveLog();
  }

  // @wire functions run around the same time as connectedCallback(), but there's no guaranteed order
  // This result in some logging requiring truthy checks using the safe navigator "?.", and some loss of logging data could occur
  @wire(returnSomeString)
  wiredReturnSomeString({ error, data }) {
    this.logger?.info('>>> logging inside a wire function, if the logger is loaded');
    if (data) {
      this.logger?.info('>>> wire function return value: ' + data);
    }
    if (error) {
      this.logger?.error('>>> wire function error: ' + JSON.stringify(error));
    }
  }
}
```

### Example: New `getLogger()` Usage

Now, `await` is no longer needed, and a `logger` instance can be immediately initialized & used. This results in less code, and happier developers 🥳 

```javascript
import { getLogger } from 'c/logger';

export default class LoggerDemo extends LightningElement {
  logger = getLogger();

  connectedCallback() {
    // Immediately use this.logger - no await, no truthy checks
    this.logger.info('Hello, world');
    this.logger.saveLog();
  }
}

  @wire(returnSomeString)
  wiredReturnSomeString({ error, data }) {
    // Immediately use this.logger - no await, no truthy checks
    this.logger.info('>>> logging inside a wire function');
    if (data) {
      this.logger.info('>>> wire function return value: ' + data);
    }
    if (error) {
      this.logger.error('>>> wire function error: ' + JSON.stringify(error));
    }
  }
}
```

## New `exception()` function

Resolved #763 by adding a JS function equivalent to the Apex method `Logger.exception()`. Both of these do 3 things in 1 line of code:
1. Log an exception
2. Save the log
3. Rethrow the exception

Previously in JS, this would have been 3 lines of code:

```javascript
this.logger.error('An error occurred').setExceptionDetails(someJSError);
this.logger.saveLog();
throw someJSError;
```

Now, 1 line of code provides the same functionality:

```javascript
this.logger.exception('An error occurred', someJSError);
```

## More Improvements for JS Stack Trace Parsing

Fixed #776 by updating logic in `loggerStackTrace.js` to better handle parsing when [lightning debug mode](https://help.salesforce.com/s/articleView?id=sf.aura_debug_mode.htm&language=en_US&type=5) is disabled
Previously, stack traces worked when debug mode was enabled, but was still inaccurate in some browsers when debug mode was off due to some subtle differences in the generated stack traces.

# Installation Info

## Core Unlocked Package - no namespace

Full Changelog: https://github.com/jongpie/NebulaLogger/compare/v4.14.12...v4.14.13

- SF CLI: `sf package install --wait 20 --security-type AdminsOnly --package 04t5Y0000015oW3QAI`
- SFDX CLI: `sfdx force:package:install --wait 20 --securitytype AdminsOnly --package 04t5Y0000015oW3QAI`
- Sandbox: https://test.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000015oW3QAI
- Production: https://login.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000015oW3QAI

<hr /><em>This discussion was created from the release <a href='https://github.com/jongpie/NebulaLogger/releases/tag/v4.14.13'>New synchronous JavaScript function getLogger() + deprecated async function createLogger() </a>.</em>

**Comment by jamessimone** [comment]
Awesome stuff!

**Reply by jongpie** [reply]
Awww yeah! Thanks again for all your help!

**Comment by chazwatkins** [comment]
@jongpie You do a great job getting releases out so often adding new features and supporting issues.  There are so many releases between NebulaLogger and ApexRollup that I feel like I need to set up a pipeline to upgrade them regularly instead of doing it manually.

**Reply by chazwatkins** [reply]
Also, I love the new way to instantiate Logger in JS.  I had been following an old way of putting the element in the html and using `this.template.querySelector` to get it.

**Reply by jongpie** [reply]
Thanks @chazwatkins! If you do happen to set up something to auto-upgrade Nebula Logger & Apex Rollup in your pipeline, let us know - I'd be very interested to see how you get it working (if you decide to do that), I'm sure @jamessimone would love to see it too.

And yeah, if you're still using the `this.template.querySelector` approach, I think the new `getLogger()` import function will be much simpler & nicer to use. (But let me know if you run into any issues with it 😅 )
